[cuda] introduce trace tracker callback in cache allocator

minsii opened the pull request:
Summary:
This patch prototypes a trace tracker callback mechanism based on existing TraceEntry records.

- It allows external of cache allocator to "attach" trace tracker callbacks. 
- When a TraceEntry is recorded, it triggers all attached callbacks. Callbacks can selectively behave based on the trace action. 
- **RISK**: The attached callback would be called within an allocator call stack (e.g., free during an allocate call). Potential deadlock may occur if other locks are called within the callback and has interdependency w/ the device allocator lock. It is the callback developer's responsibility to avoid any potential deadlock. 
- **ADVICE**: The callback mechanism is designed **only for Pytorch internal use**. We should not expose it to Python layer due to Python GIL that would cause a deadlock.

See example in D50726970 that attaches NCCL register/deregister hooks via the trace tracker callback, so that all CUDA segments allocated by the allocator can be registered to NCCL communicators before any NCCL communication happens. This enables fast zero copy algorithms in NCCL.

Differential Revision: D50726971

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -21,6 +21,7 @@
 
 #include <c10/util/Exception.h>
 #include <cuda_runtime_api.h>
+#include <folly/Utility.h>
```
pytorch doesn't take a folly dependency

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -884,6 +885,11 @@ class DeviceCachingAllocator {
+
+  // enable record trace when any trace tracker has been attached to the
+  // allocator
+  bool enable_trace_tracker_ = false;
```
This can just be replaced with `trace_trackers_.size() > 0`, same number of instructions to check.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -969,6 +977,11 @@ class DeviceCachingAllocator {
   }
 
+  void attachAllocatorTraceTracker(AllocatorTraceTracker tracker) {
+    enable_trace_tracker_ = true;
```
This needs to take the lock before mutating DeviceCachingAllocator state.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -1053,12 +1066,13 @@ class DeviceCachingAllocator {
       std::string proc_info = reportProcessMemoryInfo(device);
 
-      if (record_history) {
+      if (record_history || enable_trace_tracker_) {
```
Now the condition for calling `record_trace` is more than a single bool, we should refactor to move the check into the `record_trace` function. The other stuff that is done within these blocks (e.g. block->context_when_allocated = std::move(context)`) is cheap enough to just run universally.

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -157,16 +157,19 @@ struct TraceEntry {
   int64_t addr_; // for OOM, this is the amount of free bytes reported by cuda
   std::shared_ptr<GatheredContext> context_;
   cudaStream_t stream_;
+  int device_;
```
Since this is 32-bits, we should place this right after `Action` (also 32-bits), so that the size of struct doesn't get bigger.

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -246,6 +251,9 @@ class CUDAAllocator : public Allocator {
       size_t alloc_trace_max_entries,
       RecordContext when) = 0;
   virtual void attachOutOfMemoryObserver(OutOfMemoryObserver observer) = 0;
+  // We require AllocatorTraceTracker callback never include Pytorch call to
```
//`AllocatorTraceTracker` will be called while the per-device allocator lock is held. Any additional locks taken from within the callback must be proven to always have the lock order <allocator lock> -> <my lock>, otherwise deadlocks will occur. In particular, Python's GIL may be held when calling the allocator so it is unsafe to try to acquire the GIL in this callback.

minsii replied:
Thank you for the detailed restriction description! I will use it in the comment here.

zdevito reviewed (commented):
This looks good, I only have minor comments. It will need some test from within PyTorch to make sure the events get called.

minsii commented:
Addressed comments and added a unit test

minsii commented:
@pytorchbot label "topic: not user facing"

zdevito reviewed (approved):
Looks good! Thanks for getting that test to run

minsii committed:
[prototype][cuda] introduce trace tracker callback in cache allocator (#112238)

Summary:

This patch prototypes a trace tracker callback mechanism based on existing TraceEntry records.

- It allows external of cache allocator to "attach" trace tracker callbacks.
- When a TraceEntry is recorded, it triggers all attached callbacks. Callbacks can selectively behave based on the trace action.
- **RISK**: The attached callback would be called within an allocator call stack (e.g., free during an allocate call). Potential deadlock may occur if other locks are called within the callback and has interdependency w/ the device allocator lock. It is the callback developer's responsibility to avoid any potential deadlock.
- **ADVICE**: The callback mechanism is designed **only for Pytorch internal use**. We should not expose it to Python layer due to Python GIL that would cause a deadlock.

See example in D50726970 that attaches NCCL register/deregister hooks via the trace tracker callback, so that all CUDA segments allocated by the allocator can be registered to NCCL communicators before any NCCL communication happens. This enables fast zero copy algorithms in NCCL.

Reviewed By: zdevito

Differential Revision: D50726971
